### PR TITLE
Force load archives with Swift symbols

### LINKED ISSUE #16: zld is missing the force_load fix in Xcode 11
Xcode 10.2 introduced a linker bug that caused linking Swift in static libraries to fail:

https://developer.apple.com/documentation/xcode_release_notes/xcode_10_2_release_notes/swift_5_release_notes_for_xcode_10_2?language=objc

> Linking against a static Swift library might create a binary with missing type metadata because the object files that define the metadata inside the static archive are mistakenly considered unused. (47598583)
> 
> This can manifest as a Swift runtime error with a message such as: “failed to demangle superclass of MyClass from mangled name ‘<mangled name>’”.
> 
> Workaround: If you can rebuild the static library, try building it with whole module optimization enabled. Otherwise, add -all_load to the linker flags in the client binary to ensure all object files are linked into it.

This was fixed in Xcode 11:
https://developer.apple.com/documentation/xcode_release_notes/xcode_11_release_notes?language=objc

> Static libraries are now always force-loaded in their entirety during linking, fixing most “unable to demangle” runtime errors. (47598583)

We will need to reimplement this fix. My suspicion is the solution was not as simple as described, that they in fact only added force_load for static libraries that contained Swift symbols. A quote from the [Swift bug ticket](https://bugs.swift.org/browse/SR-10721) has:

> You should be safe if you force load any .a file that contains symbols matching the regex ^_?\$s.*Mn$ (that is, beginning with the prefix $s or _$s, and ending with the suffix Mn)

### PR DESCRIPTION
**Pull request by rmaz**
This is a fix for radar 47598583:

"Linking against a static Swift library might create a binary with missing type metadata because the object files that define the metadata inside the static archive are mistakenly considered unused. (47598583)"

The issue was fixed in ld64 in Xcode 11:

"Static libraries are now always force-loaded in their entirety during linking, fixing most “unable to demangle” runtime errors. (47598583)"

The PR applies -force_load to all archives containing any Swift symbols. Fixes #16, #20.

### CODE REVIEW

**Comment by rmaz** [comment]
I initially added this as part of the `-ObjC` logic to do per object file loading, but it caused an obscure issue with one of our test classes. I take this to mean that there can be object files in archives relied on by the Swift runtime that have no Swift symbols in them.

File: ld/src/ld/parsers/archive_file.cpp [hunk 1]
```diff
@@ -536,6 +537,17 @@ bool File<A>::forEachAtom(ld::File::AtomHandler& handler) const
 	return didSome;
 }
 
+template <typename A>
+bool File<A>::hasSwift() const {
+	for (const auto& entry : _hashTable) {
```
Comment (michaeleisel) [inline ld/src/ld/parsers/archive_file.cpp]: what perf cost do you see for this?

Comment (michaeleisel) [reply ld/src/ld/parsers/archive_file.cpp]: e.g., if you add a CACurrentMediaTime at the start and end of this function, and then sum up all the times, what % of total time is it?

Comment (rmaz) [reply ld/src/ld/parsers/archive_file.cpp]: The cost is within measurement noise for our codebase. I profiled linking our largest app before and after applying this PR, I got average link times of 38.37s vs 39.24s with a variance of > 1s for each run.

Comment (michaeleisel) [reply ld/src/ld/parsers/archive_file.cpp]: fair enough

**Pull request merged**